src/Development/IDE/Core/Shake.hs

```diff
@@ -379,57 +403,126 @@ shakeProfile :: IdeState -> FilePath -> IO ()
+        q <- newTQueue
+        traverse_ (writeTQueue q) userActs
+        return q
+    actionInProgress :: TVar (Maybe (Action())) <- newTVarIO Nothing
+
+    let
+        -- A daemon-like action used to inject additional work
+        -- Runs actions from the work queue sequentially
+        pumpAction =
+            forever $ do
+                join $ liftIO $ atomically $ do
+                    act <- readTQueue actionQueue
+                    writeTVar actionInProgress $ Just act
+                    return act
+                liftIO $ atomically $ writeTVar actionInProgress Nothing
+
+        progressRun
+          | reportProgress = lspShakeProgress ideTesting getLspId eventer inProgress
+          | otherwise = return ()
+
+        workRun restore = withAsync progressRun $ \progressThread -> do
+          let systemActs' =
+                [ [] <$ pumpAction
+                , parallel systemActs
+                    -- Only system actions are considered for progress reporting
+                    -- When done, cancel the progressThread to indicate completion
+                    <* liftIO (cancel progressThread)
+                ]
+          res <- try @SomeException
+                 (restore $ shakeRunDatabaseProfile shakeProfileDir shakeDb systemActs')
+          let res' = case res of
+                      Left e -> "exception: " <> displayException e
+                      Right _ -> "completed"
+              profile = case res of
+                      Right (_, Just fp) ->
+                          let link = case filePathToUri' $ toNormalizedFilePath' fp of
+                                          NormalizedUri _ x -> x
+                          in ", profile saved at " <> T.unpack link
+                      _ -> ""
+              -- Wrap up in a thread to avoid calling interruptible
+              -- operations inside the masked section
+          let wrapUp = logDebug logger $ T.pack $ "Finishing build session(" ++ res' ++ profile ++ ")"
+          return wrapUp
+
+    -- Do the work in a background thread
+    workThread <- asyncWithUnmask workRun
+
+    -- run the wrap up unmasked
+    _ <- async $ join $ wait workThread
+
+    -- 'runInShakeSession' is used to append work in this Shake session
+    --  The session stays open until 'cancelShakeSession' is called
+    let runInShakeSession :: forall a . Action a -> IO (IO a)
+        runInShakeSession act = do
+              res <- newBarrier
+              let act' = actionCatch @SomeException (Right <$> act) (pure . Left)
+              atomically $ writeTQueue actionQueue (act' >>= liftIO . signalBarrier res)
+              return (waitBarrier res >>= either throwIO return)
+
+    --  Cancelling is required to flush the Shake database when either
+    --  the filesystem or the Ghc configuration have changed
+        cancelShakeSession = do
+            cancel workThread
+            atomically $ do
+                q <- flushTQueue actionQueue
```

Comment:
Can you explain what is going on here? How are things requeued?

Reply:
`cancelShakeSession` returns a list of the pending tasks, which are then fed back into the new session in line 438